vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+                container_id: req.id.to_string(),
+                io_uid: self.io_uid,
+                io_gid: self.io_gid,
+                spec: p,
+                exit_signal: Default::default(),
+            }),
+            stdin: Arc::new(Mutex::new(None)),
+        })
+    }
+}
+
+pub struct YoukiInitLifecycle {
+    youki_container: Arc<Mutex<Container>>,
+    exit_signal: Arc<ExitSignal>,
+}
+
+#[async_trait]
+impl ProcessLifecycle<InitProcess> for YoukiInitLifecycle {
+    async fn start(&self, p: &mut InitProcess) -> containerd_shim::Result<()> {
+        p.lifecycle
+            .youki_container
+            .lock()
+            .await
+            .start()
+            .map_err(other_error!(e, "failed to start container "))?;
+        p.state = Status::RUNNING;
+        Ok(())
+    }
+
+    async fn kill(
+        &self,
+        p: &mut InitProcess,
+        signal: u32,
+        all: bool,
+    ) -> containerd_shim::Result<()> {
+        let signal = Signal::try_from(signal as i32)
+            .map_err(other_error!(e, "failed to parse kill signal "))?;
+        p.lifecycle
+            .youki_container
+            .lock()
+            .await
+            .kill(signal, all)
+            .or_else(|e| {
+                if let LibcontainerError::IncorrectStatus = e {
+                    Ok(())
+                } else {
+                    Err(e)
+                }
+            })
+            .map_err(other_error!(e, "failed to kill container "))?;
+        Ok(())
+    }
+
+    async fn delete(&self, p: &mut InitProcess) -> containerd_shim::Result<()> {
+        p.lifecycle
+            .youki_container
+            .lock()
+            .await
+            .delete(true)
+            .map_err(other_error!(e, "failed to delete container "))?;
+        self.exit_signal.signal();
+        Ok(())
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn update(&self, p: &mut InitProcess, resources: &LinuxResources) -> Result<()> {
+        if p.pid <= 0 {
+            return Err(other!(
+                "failed to update resources because init process is {}",
+                p.pid
+            ));
+        }
+        containerd_shim::cgroup::update_resources(p.pid as u32, resources)
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    async fn update(&self, _p: &mut InitProcess, _resources: &LinuxResources) -> Result<()> {
+        Err(Error::Unimplemented("update resource".to_string()))
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn stats(&self, p: &InitProcess) -> Result<Metrics> {
+        if p.pid <= 0 {
+            return Err(other!(
+                "failed to collect metrics because init process is {}",
+                p.pid
+            ));
+        }
+        containerd_shim::cgroup::collect_metrics(p.pid as u32)
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    async fn stats(&self, _p: &InitProcess) -> Result<Metrics> {
+        Err(Error::Unimplemented("process stats".to_string()))
+    }
```

Comment:
Same as above

Reply:
I think it is originally from the rust-extensions